### ISSUE TITLE
lingua or touchtune or torchtitan?

### ISSUE THREAD

**Issue by youngsheen** [facebookresearch/lingua#13]
I’m a bit confused about the differences between Lingua, TouchTune, and TorchTitan. From my understanding, TouchTune is used for SFT, while TorchTitan is intended for pretraining. Could someone clarify the role of Lingua and in what scenarios it should be used for development?

**Comment by DeiseFreire** [comment]
Hello @youngsheen , language refers to all languages. Torch Titan creates the knowledge base. TouchTune adapts the knowledge to specific tasks. I hope this helps you.

**Comment by pbontrager** [comment]
I can chime in here too as a torchtune maintainer. There isn’t a cut and dry answer your question. Each of these three libraries have been designed with a different user in mind but there’s also a lot of overlap so each user can choose what they prefer. 
- torchtitan demonstrates how to do pre-training at scale and how to take advantage of all the scaling techniques for maximum performance, but doesn’t focus on experimentation of model design
- lingua is a pre-training library with a strong focus on hackability and experimentation
- torchtune has a similar ethos to lingua focusing on hackability and minimal abstractions but its geared more for finetuning based on the integrations we offer and by focusing more on efficient memory usage and lower gpu resources.

A developer could probably use any of these libraries for all of their language modeling tasks but with different trade offs.

**Comment by BadrYoubiIdrissi** [comment]
Hi ! Thank you for the very relevant question ! We meant to add a section on this for the readme but it didn't make it in time :) 

Lingua is designed for researchers who want to experiment with new ideas for LLM pretraining and get quick feedback on both training/inference speed and downstream benchmarks. Our goal is to lower the barrier to entry for LLM research by providing a lightweight and focused codebase.

We see torchtitan, torchtune, and lingua as complementary tools. Torchtitan is excellent for large-scale work because it features 3D parallelism and is likely to integrate the latest PyTorch distributed training features more quickly, thanks to its close ties to the PyTorch team. On the other hand, Torchtune excels at fine-tuning, especially when GPU resources are limited, by offering various fine-tuning strategies like LoRA, QLoRA, DPO, and PPO.

A typical workflow could look like this: you might first test a new idea in Lingua, then scale it up further with Torchtitan, and finally use Torchtune for instruction or preference fine-tuning.

Although there's definitely some overlap among these codebases, we think it's valuable to have focused tools for different aspects of LLM work. For example, Torchtitan aims to showcase the latest distributed training features of PyTorch in a clean, minimal codebase, but for most research, you really don't need every feature PyTorch has to offer or the capability to scale to 100B parameters on 4096 GPUs. For instance, we think that FSDP + torch compile will cover 90% of all needs of a researcher. With lingua, we tried to ask "What's the minimal set of features needed to draw solid conclusions on the scalability of idea X?"

We believe this targeted approach helps researchers make progress faster without the mental overhead of using many techniques that might not be needed.

**Comment by DeiseFreire** [comment]
@BadrYoubiIdrissi Oh, very good. I appreciate that!

> _"efficient memory usage and lower gpu resources."_

**Comment by DeiseFreire** [comment]
@BadrYoubiIdrissi Yes, that sentence alone won me over:

> "We believe this targeted approach helps researchers make progress faster without the mental overhead of using many techniques that might not be needed."
